ledger/participant-integration-api/src/main/scala/platform/indexer/parallel/ParallelIndexerFactory.scala

```diff
@@ -115,16 +98,160 @@ object ParallelIndexerFactory {
+    )
+  }
+
+  def seqMapperZero(initialSeqId: Long): RunningBatch[Vector[DBDTOV1]] =
+    RunningBatch(
+      lastOffset = null,
+      lastSeqEventId = initialSeqId, // this is the only property of interest in the zero element
+      lastConfig = None,
+      batch = Vector.empty,
+      batchSize = 0,
+      averageStartTime = 0,
+    )
+
+  def seqMapper(metrics: Metrics)(
+      previous: RunningBatch[Vector[DBDTOV1]],
+      current: RunningBatch[Vector[DBDTOV1]],
+  ): RunningBatch[Vector[DBDTOV1]] = {
+    var eventSeqId = previous.lastSeqEventId
+    val batchWithSeqIds = current.batch.map {
+      case dbDto: DBDTOV1.EventCreate =>
+        eventSeqId += 1
+        dbDto.copy(event_sequential_id = eventSeqId)
+
+      case dbDto: DBDTOV1.EventExercise =>
+        eventSeqId += 1
+        dbDto.copy(event_sequential_id = eventSeqId)
+
+      case dbDto: DBDTOV1.EventDivulgence =>
+        eventSeqId += 1
+        dbDto.copy(event_sequential_id = eventSeqId)
+
+      case notEvent => notEvent
+    }
+    val nowNanos = System.nanoTime()
+    metrics.daml.parallelIndexer.inputMapping.duration.update(
+      (nowNanos - current.averageStartTime) / current.batchSize,
+      TimeUnit.NANOSECONDS,
+    )
+    current.copy(
+      lastSeqEventId = eventSeqId,
+      batch = batchWithSeqIds,
+      averageStartTime = nowNanos, // setting start time to the start of the next stage
+    )
+  }
+
+  def batcher[DB_BATCH](
+      batchF: Vector[DBDTOV1] => DB_BATCH,
+      metrics: Metrics,
+  ): RunningBatch[Vector[DBDTOV1]] => RunningBatch[DB_BATCH] = { inBatch =>
+    val dbBatch = batchF(inBatch.batch)
+    val nowNanos = System.nanoTime()
+    metrics.daml.parallelIndexer.batching.duration.update(
+      (nowNanos - inBatch.averageStartTime) / inBatch.batchSize,
+      TimeUnit.NANOSECONDS,
+    )
+    inBatch.copy(
+      batch = dbBatch,
+      averageStartTime = nowNanos,
+    )
+  }
+
+  def ingester[DB_BATCH](
+      ingestFunction: (Connection, DB_BATCH) => Unit,
+      dbDispatcher: DbDispatcher,
+      metrics: Metrics,
+  )(implicit
+      loggingContext: LoggingContext // TODO is that right so? shouldn't we prep something here?
+  ): RunningBatch[DB_BATCH] => Future[RunningBatch[DB_BATCH]] =
+    batch =>
+      dbDispatcher.executeSql(metrics.daml.parallelIndexer.ingestion) { connection =>
+        metrics.daml.parallelIndexer.updates.inc(batch.batchSize.toLong)
+        ingestFunction(connection, batch.batch)
+        val nowNanos = System.nanoTime()
+        metrics.daml.parallelIndexer.ingestion.duration.update(
+          (nowNanos - batch.averageStartTime) / batch.batchSize,
+          TimeUnit.NANOSECONDS,
+        )
+        batch
+      }
+
+  def tailer[DB_BATCH]: (RunningBatch[DB_BATCH], RunningBatch[DB_BATCH]) => RunningBatch[DB_BATCH] =
+    (prev, curr) =>
+      RunningBatch[DB_BATCH](
+        lastOffset = curr.lastOffset,
+        lastSeqEventId = curr.lastSeqEventId,
```

Comment:
As noted above, is this correct? Can't `prev.lastSeqEventId` be greater than `curr.lastSeqEventId`?